File: packages/unity-csharp/Runtime/Wallet/DeepLinkHandler.cs
using System;
using System.Threading.Tasks;
using UnityEngine;

namespace Cinacoin
{
    /// <summary>
    /// Deep link handler for mobile builds (iOS/Android) and desktop.
    /// Opens wallet apps via deep links with fallback to universal links.
    /// Supports WalletConnect deep link protocol for wallet-to-app communication.
    /// </summary>
    public class DeepLinkHandler
    {
        /// Platform detection.
        public enum Platform
        {
            iOS,
            Android,
            WebGL,
            Web
        }

        /// Result of a redirect attempt.
        public class RedirectResult
        {
            public bool Success;
            public string Method; // deep-link | universal-link | app-store | qr-code
            public string Url;
            public bool FallbackUsed;
            public string Error;
        }

        private Platform _platform;

        public DeepLinkHandler()
        {
#if UNITY_IOS
            _platform = Platform.iOS;
#elif UNITY_ANDROID
            _platform = Platform.Android;
#elif UNITY_WEBGL
            _platform = Platform.WebGL;
#else
            _platform = Platform.Web;
#endif
        }

        /// Generate a deep link URL for a wallet.
        public string GenerateDeepLink(string walletId, string uri)
        {
            var wallet = WalletRegistry.Get(walletId);
            if (wallet == null)
                throw new ArgumentException($"Unknown wallet ID: {walletId}");

            var scheme = string.IsNullOrEmpty(wallet.DeepLinkScheme) ? "wc://" : wallet.DeepLinkScheme;

            // WalletConnect deep link format
            return $"{scheme}wc?uri={Uri.EscapeDataString(uri)}";
        }

        /// Generate a universal link URL for fallback.
        public string GenerateUniversalLink(string walletId, string uri)
        {
            var wallet = WalletRegistry.Get(walletId);
            if (wallet == null)
                throw new ArgumentException($"Unknown wallet ID: {walletId}");

            var domain = string.IsNullOrEmpty(wallet.UniversalLinkDomain)
                ? "walletconnect.com"
                : wallet.UniversalLinkDomain;

            return $"https://{domain}/wc?uri={Uri.EscapeDataString(uri)}";
        }

        /// Generate an app store URL for a wallet.
        public string GenerateAppStoreUrl(string walletId)
        {
            var wallet = WalletRegistry.Get(walletId);
            if (wallet == null)
                throw new ArgumentException($"Unknown wallet ID: {walletId}");

#if UNITY_IOS
            return wallet.AppStoreUrl;
#elif UNITY_ANDROID
            return wallet.PlayStoreUrl;
#else
            return wallet.AppStoreUrl; // Default to App Store
#endif
        }

        /// Open a deep link for a wallet app.
        /// Returns success/failure with the method used.
        public async Task<RedirectResult> OpenDeepLinkAsync(string walletId, string uri)
        {
            var wallet = WalletRegistry.Get(walletId);
            if (wallet == null)
            {
                return new RedirectResult
                {
                    Success = false,
                    Method = "deep-link",
                    Error = $"Unknown wallet ID: {walletId}"
                };
            }

            try
            {
                // On mobile: try deep link, then fallback to universal link, then app store
                // On desktop/web: just return the URI for QR display

                if (_platform == Platform.WebGL || _platform == Platform.Web)
                {
                    // Desktop/Web: cannot deep link, signal QR code display
                    return new RedirectResult
                    {
                        Success = false,
                        Method = "qr-code",
                        Url = uri,
                        Error = "QR code display required on desktop"
                    };
                }

                // Try deep link first
                var deepLinkUrl = GenerateDeepLink(walletId, uri);
                bool opened = OpenURL(deepLinkUrl);

                if (opened)
                {
                    return new RedirectResult
                    {
                        Success = true,
                        Method = "deep-link",
                        Url = deepLinkUrl
                    };
                }

                // Fallback to universal link
                var universalUrl = GenerateUniversalLink(walletId, uri);
                OpenURL(universalUrl);

                return new RedirectResult
                {
                    Success = true,
                    Method = "universal-link",
                    Url = universalUrl,
                    FallbackUsed = true
                };
            }
            catch (Exception ex)
            {
                return new RedirectResult
                {
                    Success = false,
                    Method = "qr-code",
                    Error = ex.Message
                };
            }
        }

        /// Open a URL using the platform-appropriate method.
        private bool OpenURL(string url)
        {
            if (string.IsNullOrEmpty(url)) return false;

            try
            {
                Application.OpenURL(url);
                return true;
            }
            catch
            {
                return false;
            }
        }

        /// Open the app store page for a wallet.
        public RedirectResult OpenAppStore(string walletId)
        {
            try
            {
                var url = GenerateAppStoreUrl(walletId);
                if (string.IsNullOrEmpty(url))
                {
                    return new RedirectResult
                    {
                        Success = false,
                        Method = "app-store",
                        Error = "No app store URL available"
                    };
                }

                OpenURL(url);
                return new RedirectResult
                {
                    Success = true,
                    Method = "app-store",
                    Url = url
                };
            }
            catch (Exception ex)
            {
                return new RedirectResult
                {
                    Success = false,
                    Method = "app-store",
                    Error = ex.Message
                };
            }
        }

        /// Set the target platform (for testing).
        public void SetPlatform(Platform platform)
        {
            _platform = platform;
        }

        /// Get the current platform.
        public Platform GetCurrentPlatform() => _platform;

        /// Check if deep linking is supported on the current platform.
        public bool IsDeepLinkSupported()
        {
            return _platform == Platform.iOS || _platform == Platform.Android;
        }

        /// Build a WalletConnect v2 deep link callback URL for mobile wallets.
        /// After the wallet approves, it will redirect back to this URL with the session data.
        public string BuildCallbackUrl(string protocol = "cinacoin")
        {
            return $"{protocol}://wc-callback";
        }
    }
}

File: packages/flutter-dart/lib/src/deep_link_handler.dart
/// Deep link + App Links + Universal Links handler for Flutter.
///
/// Mirrors the core-sdk links module functionality.
import 'package:app_links/app_links.dart';
import 'package:url_launcher/url_launcher.dart';
import 'types.dart';
import 'wallet_registry.dart';

/// Platform detection for redirect decisions.
enum DeepLinkPlatform {
  ios,
  android,
  web,
}

/// Parameters for generating a deep link.
class DeepLinkParams {
  final String walletId;
  final String uri;
  final Map<String, String>? queryParams;
  final int fallbackTimeoutMs;

  const DeepLinkParams({
    required this.walletId,
    required this.uri,
    this.queryParams,
    this.fallbackTimeoutMs = 1500,
  });
}

/// Result of a redirect attempt.
class RedirectResult {
  final bool success;
  final String method; // 'deep-link' | 'universal-link' | 'app-store' | 'qr-code'
  final String url;
  final bool fallbackUsed;
  final String? error;

  const RedirectResult({
    required this.success,
    required this.method,
    required this.url,
    this.fallbackUsed = false,
    this.error,
  });
}

/// Deep link handler for wallet connections.
class DeepLinkHandler {
  final AppLinks _appLinks;
  DeepLinkPlatform? _platform;

  DeepLinkHandler({AppLinks? appLinks})
      : _appLinks = appLinks ?? AppLinks();

  /// Detect the current platform.
  static Future<DeepLinkPlatform> detectPlatform() async {
    // In Flutter, we need platform-specific detection.
    // For now, return Android by default; the platform plugin
    // would typically be used for actual detection.
    return DeepLinkPlatform.android;
  }

  /// Set the target platform for redirect handling.
  void setPlatform(DeepLinkPlatform platform) {
    _platform = platform;
  }

  /// Generate a deep link URL for a specific wallet.
  ///
  /// Returns the complete deep link URL string.
  String generateDeepLink(DeepLinkParams params) {
    final wallet = WalletRegistry.get(params.walletId);
    if (wallet == null) {
      throw ArgumentError('Unknown wallet ID: ${params.walletId}. Register it in WalletRegistry.');
    }

    final scheme = wallet.deepLinkScheme ?? 'wc://';
    String url = '$scheme';

    // Apply path template
    final pathTemplate = '/wc?uri={uri}';
    url += pathTemplate.replaceFirst('{uri}', Uri.encodeComponent(params.uri));

    // Append additional query parameters
    if (params.queryParams != null && params.queryParams!.isNotEmpty) {
      final separator = url.contains('?') ? '&' : '?';
      final qs = params.queryParams!.entries
          .map((e) => '${Uri.encodeComponent(e.key)}=${Uri.encodeComponent(e.value)}')
          .join('&');
      url += separator + qs;
    }

    return url;
  }

  /// Generate a universal link URL for fallback.
  String generateUniversalLink(DeepLinkParams params) {
    final wallet = WalletRegistry.get(params.walletId);
    if (wallet == null) {
      throw ArgumentError('Unknown wallet ID: ${params.walletId}');
    }

    final domain = wallet.universalLinkDomain ?? 'walletconnect.com';
    return 'https://$domain/wc?uri=${Uri.encodeComponent(params.uri)}';
  }

  /// Open a deep link URL.
  ///
  /// Uses smart redirect: deep link → timeout → universal link → app store.
  Future<RedirectResult> openDeepLink(
    DeepLinkParams params, {
    DeepLinkPlatform? platform,
  }) async {
    final targetPlatform = platform ?? _platform ?? await detectPlatform();
    final wallet = WalletRegistry.get(params.walletId);

    if (wallet == null) {
      return RedirectResult(
        success: false,
        method: 'deep-link',
        url: '',
        error: 'Unknown wallet ID: ${params.walletId}',
      );
    }

    try {
      // Try deep link first
      final deepLinkUrl = generateDeepLink(params);
      final uri = Uri.parse(deepLinkUrl);

      final canLaunch = await canLaunchUrl(uri);
      if (canLaunch) {
        await launchUrl(uri, mode: LaunchMode.externalApplication);
        return RedirectResult(
          success: true,
          method: 'deep-link',
          url: deepLinkUrl,
        );
      }

      // Fallback to universal link
      final universalUrl = generateUniversalLink(params);
      final universalUri = Uri.parse(universalUrl);

      if (await canLaunchUrl(universalUri)) {
        await launchUrl(universalUri, mode: LaunchMode.externalApplication);
        return RedirectResult(
          success: true,
          method: 'universal-link',
          url: universalUrl,
          fallbackUsed: true,
        );
      }

      // Fallback to app store
      final storeUrl = targetPlatform == DeepLinkPlatform.ios
          ? wallet.appStoreUrl
          : wallet.playStoreUrl;

      if (storeUrl != null) {
        final storeUri = Uri.parse(storeUrl);
        if (await canLaunchUrl(storeUri)) {
          await launchUrl(storeUri, mode: LaunchMode.externalApplication);
          return RedirectResult(
            success: true,
            method: 'app-store',
            url: storeUrl,
            fallbackUsed: true,
          );
        }
      }

      return RedirectResult(
        success: false,
        method: 'qr-code',
        url: deepLinkUrl,
        error: 'Could not open wallet app. Please scan QR code.',
      );
    } catch (e) {
      return RedirectResult(
        success: false,
        method: 'deep-link',
        url: '',
        error: e.toString(),
      );
    }
  }

  /// Listen for incoming deep links (App Links / Universal Links).
  ///
  /// Returns a stream of URI strings received via deep link.
  Stream<String> listenForIncomingLinks() async* {
    await for (final uri in _appLinks.uriLinkStream) {
      yield uri.toString();
    }
  }

  /// Get initial URI if the app was launched via deep link.
  Future<Uri?> getInitialLink() async {
    try {
      return await _appLinks.getInitialLink();
    } catch (_) {
      return null;
    }
  }
}
